akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphWireTapSpec.scala

```diff
@@ -0,0 +1,71 @@
+package akka.stream.scaladsl
+
+import akka.stream._
+import akka.stream.testkit.Utils._
+import akka.stream.testkit._
+import akka.stream.testkit.scaladsl.TestSink
+
+class GraphWireTapSpec extends StreamSpec {
+
+  val settings = ActorMaterializerSettings(system)
+    .withInputBuffer(initialSize = 2, maxSize = 16)
+
+  implicit val materializer = ActorMaterializer(settings)
+
+  "A wire tap" must {
+
+    "wireTap must broadcast to the tap" in assertAllStagesStopped {
+      val tp, mp = TestSink.probe[Int](system)
+      val (tps, mps) = Source(1 to 2).wireTapMat(tp)(Keep.right).toMat(mp)(Keep.both).run()
+      tps.request(2)
+      mps.requestNext(1)
+      mps.requestNext(2)
+      tps.expectNext(1, 2)
+      mps.expectComplete()
+      tps.expectComplete()
+    }
+
+    "wireTap must drop elements while the tap has no demand, buffering up to one element" in assertAllStagesStopped {
+      val tp, mp = TestSink.probe[Int](system)
+      val (tps, mps) = Source(1 to 6).wireTapMat(tp)(Keep.right).toMat(mp)(Keep.both).run()
+      mps.request(3)
+      mps.expectNext(1, 2, 3)
+      tps.request(4)
+      mps.requestNext(4)
+      mps.requestNext(5)
+      mps.requestNext(6)
+      tps.expectNext(3, 4, 5, 6)
+      mps.expectComplete()
+      tps.expectComplete()
+    }
+
+    "wireTap must broadcast all elems to the tap if it has a large enough buffer" in assertAllStagesStopped {
```

Comment:
Is this a good/necessary test? Is it guaranteed to pass, and also, _ought_ it be guaranteed to pass if the `WireTap` is not to have surprising behaviour?

Reply:
I think so, at least if it is racy we hopefully will notice that at some point when the test fails :)

Reply:
Right, I've seen this fail once with:

```
assertion failed: expected OnNext(2), found OnNext(3)
java.lang.AssertionError: assertion failed: expected OnNext(2), found OnNext(3)
	at scala.Predef$.assert(Predef.scala:170)
```